convert/convert.go

```diff
@@ -318,321 +337,317 @@ func WithChunksCompression(compressionOpts ...schema.CompressionOpts) ConvertOpt
-		tsdbSchema:  s,
-		concurrency: ops.concurrency,
+			hasChunks := slices.ContainsFunc(chks, func(chk chunks.Meta) bool {
+				return mint <= chk.MaxTime && chk.MinTime <= maxt
+			})
+			if !hasChunks {
+				continue
+			}
 
-		rowBuilder: parquet.NewRowBuilder(s.Schema),
-		encoder:    schema.NewPrometheusParquetChunksEncoder(s, ops.maxSamplesPerChunk),
+			scratchBuilderLabels := scratchBuilder.Labels()
+			allSeries = append(allSeries, blockSeries{
+				blockIdx:  blockIndexReader.idx,
+				seriesIdx: i,
+				ref:       blockIndexReader.postings.At(),
+				labels:    scratchBuilderLabels,
+			})
+		}
 	}
-	ok = true
-	return rr, nil
-}
 
-func (rr *TsdbRowReader) Close() error {
-	err := &multierror.Error{}
-	for i := range rr.closers {
-		err = multierror.Append(err, rr.closers[i].Close())
+	if len(allSeries) == 0 {
+		return 0, nil, nil
 	}
-	return err.ErrorOrNil()
-}
-
-func (rr *TsdbRowReader) Schema() *schema.TSDBSchema {
-	return rr.tsdbSchema
-}
 
-func sortedPostings(ctx context.Context, indexr tsdb.IndexReader, mint, maxt int64, sortedLabels ...string) index.Postings {
-	p := tsdb.AllSortedPostings(ctx, indexr)
+	slices.SortFunc(allSeries, compareBlockSeriesBySortedLabelsFunc(opts.sortedLabels))
 
-	if len(sortedLabels) == 0 {
-		return p
-	}
-
-	type s struct {
-		ref    storage.SeriesRef
-		idx    int
-		labels labels.Labels
+	// Count how many unique series will exist after merging across blocks.
+	uniqueSeriesCount := 1
+	for i := 1; i < len(allSeries); i++ {
+		if labels.Compare(allSeries[i].labels, allSeries[i-1].labels) != 0 {
+			uniqueSeriesCount++
+		}
 	}
-	series := make([]s, 0, 128)
-	chks := make([]chunks.Meta, 0, 128)
 
-	scratchBuilder := labels.NewScratchBuilder(10)
-	lb := labels.NewBuilder(labels.EmptyLabels())
-	i := 0
-P:
-	for p.Next() {
-		scratchBuilder.Reset()
-		chks = chks[:0]
-		if err := indexr.Series(p.At(), &scratchBuilder, &chks); err != nil {
-			return index.ErrPostings(fmt.Errorf("unable to expand series: %w", err))
-		}
-		hasChunks := slices.ContainsFunc(chks, func(chk chunks.Meta) bool {
-			return mint <= chk.MaxTime && chk.MinTime <= maxt
-		})
-		if !hasChunks {
-			continue P
+	// Divide rows evenly across shards to avoid one small shard at the end;
+	// Use (a + b - 1) / b equivalence to math.Ceil(a / b)
+	// so integer division does not cut off the remainder series and to avoid floating point issues.
+	totalShards := (uniqueSeriesCount + (opts.numRowGroups * opts.rowGroupSize) - 1) / (opts.numRowGroups * opts.rowGroupSize)
+	rowsPerShard := (uniqueSeriesCount + totalShards - 1) / totalShards
+
+	// For each shard index i, shardSeries[i] is a map of blockIdx -> []series.
+	shardSeries := make([]map[int][]blockSeries, totalShards)
+	for i := range shardSeries {
+		shardSeries[i] = make(map[int][]blockSeries)
+	}
+
+	shardIdx, allSeriesIdx := 0, 0
+	for shardIdx < totalShards {
+		seriesToShard := allSeries[allSeriesIdx:]
+
+		i, uniqueCount := 0, 0
+		matchLabels := labels.Labels{}
```

Comment:
Using this as a sentinal comparator to start with assumes all series have at least one label which I am assuming must be the case, outside of the possibility of constructing a series with no `__name__` in a test